src/main/java/net/minestom/server/extras/vanilla/blocks/DAxisBlock.java

```diff
@@ -0,0 +1,71 @@
+package net.minestom.server.extras.vanilla.blocks;
+
+import net.minestom.server.data.Data;
+import net.minestom.server.entity.Player;
+import net.minestom.server.extras.vanilla.VanillaBlocks;
+import net.minestom.server.instance.Chunk;
+import net.minestom.server.instance.Instance;
+import net.minestom.server.instance.block.Block;
+import net.minestom.server.instance.block.BlockFace;
+import net.minestom.server.instance.block.CustomBlock;
+import net.minestom.server.utils.BlockPosition;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class DAxisBlock extends CustomBlock {
```

Comment:
As I said in another comment, but to add upon it:
With this API, the `AxisPlacementRule` would be unnecessary, and could be removed, Unless I am forgetting something

Reply:
This PR would force user to use custom block if they want placement rule

Reply:
Yes. Do you think that is a bad thing? If so, why?

Reply:
Yes. Placement rules should be independent from functionality. For example, if I want to create a chest behavior (right click to open an inventory) I would add a custom block for it. If I now want to add that behavior to a chest, I cannot because the chest already has a custom block to handle its placement. I could combine them but then I can't use the chest on non-chest blocks.

Reply:
Maybe I am thinking wrong but: If you wanted to add a chest with inventory, you would make a new class that extends `RChestBlock` and implement `onInteract` there. You would then use `VanillaBlocks.registerBlock` to override the vanilla implementation with your own so that when you place a chest, it will make it your chest.

Reply:
Actually, I think I failed to see what you said. I need to think about it for a bit.

Reply:
I probably said it in an odd way. There are two aspects of it.

1. If we require all blocks with a placement rule to be custom blocks then it is more to store in the world, when in reality they are only relevant when placed (or a neighbor is placed).
2. (my first point) Extending `RChestBlock` is fine, however it is limiting. It means that both the chest placement and its functionality are tied to the same object meaning I cannot reuse the chest logic for another block which may not have chest placement rules.

Finally, I am unsure on the advantage of using a custom block over a placement rule?

Reply:
I think you could do something like this if you wanted your custom block to be independent.

Your custom block would extend this: [gist](https://gist.github.com/EnanderW/6151baecbb508018227a2eaabc1a549c)
That would work right? Perhaps not super pretty, but it's way more efficient than the old way (separate placement and all that stuff).

Reply:
The advantage is performance. Right now there are a lot of unnecessary stuff happening, like looping through all blocks in the area when some blocks might not even use that at all (log, wood for example). Also, is it really more storage? I thought all blocks are stored within a long, half is custom block, half is normal. I suppose the additional custom blocks per real block is added, but that is very minimal I would say.

Reply:
I think that you could also just reuse the same custom block object for all blocks that share mechanics.

Reply:
It is pretty restricting, what if you want to use the same subclass for all your custom blocks?

It also means that blocks need to be mapped to a custom block for player placement. Which defeat the purpose of differentiating the visual from the "custom id"

Reply:
The example I sent in the gist should remove the restriction. I'd even argue that it's less restricted than the current version even without it.

I don't see how the mapping defeats the purpose though? What purpose, and is it really necessary?

Reply:
The current version is only related to the block id so less restricting, and the performance impact should be fairly minimal as blocks are not placing too frequently. The computation can even be avoided altogether using `Batch`

I am not a fan of merging multiple features together, in your case the custom block id of the blocks you registered will break will every version, and can conflict with other user blocks

Reply:
I guess I am misunderstanding something. Why would they break? Regarding the ids, that is what the constant `VANILLA_LAST_ID ` is for. I guess it's not pretty though. If you were to make your own custom block, you would set your ids to `VANILLA_LAST_ID + your_id`. This is problematic if there are different custom blocks coming from different sources (extensions for example), but that problem exists anyways.